### ISSUE TITLE
Problem connecting rasa chatbot with facebook messenger

### ISSUE THREAD

**Issue by pudasainishushant** [RasaHQ/rasa#4103]
Rasa Core version: Rasa 1.1.6
Python version: Python 3.6.8 :: Anaconda, Inc.

Operating system (windows, osx, ...): Ubuntu

Issue:  The rasa chatbot works perfectly in rasa shell.
![Screenshot from 2019-07-25 12-18-07](https://user-images.githubusercontent.com/47980452/61852145-70ca8400-aed8-11e9-9a01-c461e3d9539f.png)

 I am trying to integrate this rasa chatbot to facebook messenger. I have created a facebook page. Also developed a facebook app messenger from developers.facebook.com 
![Screenshot from 2019-07-25 12-43-41](https://user-images.githubusercontent.com/47980452/61852716-e125d500-aed9-11e9-86dd-4a8616a467c1.png)
Used the access token from here and secret from here
![Screenshot from 2019-07-25 12-44-58](https://user-images.githubusercontent.com/47980452/61852842-29dd8e00-aeda-11e9-919b-031e83877ad7.png)
Validated the webhooks through ngrok
Run rasa through this 
![Screenshot from 2019-07-25 12-47-57](https://user-images.githubusercontent.com/47980452/61853000-817bf980-aeda-11e9-9751-8f2155434f59.png)
Works perfectly on rest api
But not with facebook webhooks
![Screenshot from 2019-07-25 12-49-55](https://user-images.githubusercontent.com/47980452/61853137-dfa8dc80-aeda-11e9-9efb-d5897ed56eca.png)
![Screenshot from 2019-07-25 12-50-19](https://user-images.githubusercontent.com/47980452/61853146-e20b3680-aeda-11e9-94a6-adc6ba30a052.png)
![Screenshot from 2019-07-25 12-50-35](https://user-images.githubusercontent.com/47980452/61853151-e3d4fa00-aeda-11e9-861b-074bad899b3c.png)

Content of domain file (if used & relevant):

**Comment by MetcalfeTom** [comment]
Thanks for raising this issue, @GhostVV will get back to you about it soon!

**Comment by MetcalfeTom** [comment]
@pudasainishushant Hi, thanks for the screenshots!

Can you communicate with the Facebook graph API through POSTMAN?  e.g. with the [`/me` endpoint?](https://developers.facebook.com/docs/graph-api/using-graph-api/)

**Comment by pudasainishushant** [comment]
![Screenshot from 2019-07-28 10-47-23](https://user-images.githubusercontent.com/47980452/62002509-533e2a00-b125-11e9-89f4-6f9fef169fba.png)
This is the screenshot when i try to check it from Postman. It says not validated and wrong facebook secret but I have provided exact secret token and page access token and verify token in the credentials.yml file

**Comment by lahsuk** [comment]
Hi @pudasainishushant ,

You need to add your secret to the credentails.yml file too. Here is the link https://rasa.com/docs/rasa/user-guide/connectors/facebook-messenger/ . You need to add all verify, secret and page-access-token to the credentials file.

And only after that should you verify the webhook after forwarding port through ngrok.

Hope that helps.

**Comment by pudasainishushant** [comment]
Hello @MetcalfeTom , thanks for reminding me that. However, I have already added verify token, secret and page_access_token to the credentials file. Here, is the screenshot of that. 
![Screenshot from 2019-07-29 16-37-35](https://user-images.githubusercontent.com/47980452/62043181-72b97d80-b21f-11e9-8af5-ea73172328f9.png)

**Comment by lahsuk** [comment]
Hi @pudasainishushant ,

Did you verify your webhook? And if you did, you should test it through messenger not from postman.

**Comment by pudasainishushant** [comment]
Hello @lahsuk Yes i verfied my webhooks on facebook developers app
![Screenshot from 2019-07-25 12-43-41](https://user-images.githubusercontent.com/47980452/62186514-40766000-b386-11e9-8fa8-30d270dd1d7e.png)

**Comment by lahsuk** [comment]
Hi @pudasainishushant , 

You need to verify the webhook every time your webhook changes. It's not a one-time thing. (From what I can see, your url from ngrok changes every time)  
And like I said before, try checking if it works from messenger rather than from postman.

**Comment by pudasainishushant** [comment]
Hi, @lahsuk , I highly appreciate you trying to help me out. But, this doesn't work. When I sent GET request to my server, it works fine through ngrok. But, when I send POST request to my server, through any other facebook to the messenger bot in the developed page I get SSL POST request OK in ngrok but, I get SSL certification error on localhost:5005 where I ran rasa run .
I get this error "Exception when trying to handle webhook for facebook message." I verfied webhook from here after running ngrok. This facebook edit subsciption get verified. Here are the complete screenshots.
![Screenshot from 2019-08-01 16-29-51](https://user-images.githubusercontent.com/47980452/62287557-df2fb900-b479-11e9-8c0b-06ff6466f538.png)
![Screenshot from 2019-08-01 16-30-09](https://user-images.githubusercontent.com/47980452/62287561-e060e600-b479-11e9-99ab-e728ed838f6e.png)
![Screenshot from 2019-08-01 16-30-58](https://user-images.githubusercontent.com/47980452/62287564-e22aa980-b479-11e9-85f3-63785fec1416.png)

**Comment by pudasainishushant** [comment]
Hello, @MetcalfeTom , @lahsuk , @JustinaPetr , @sheldon , @fungiboletus , @tmbo , @amn41 , @ghostHBL , @ricwo . Can you please help me out in this issue?

**Comment by lahsuk** [comment]
@pudasainishushant 
Well, I don't know what you've done so, here is a list that you should do:
- [ ] You need to add the page in the apps in developers for fb (under messenger > settings) and additionally edit subscription to include at least messages, messaing_postbacks, messaing_read, messaging_handovers and message_delivers
- [ ] create new page access tokens and secret and add it to credentials in your rasa project
- [ ] On the facebook page settings under messenger platform, enable response to be automated and configure it to set primary receiver protocol to your developers app and secondary to inbox (or anything else)
- [ ] you need to subscribe to webhooks for developers app. (change the url and try again if you've already done it just in case)
- [ ] try testing on messenger if it works

I may have missed something but if it doesn't work, just message me again.
Hope that helps.

**Comment by pudasainishushant** [comment]
![Screenshot from 2019-08-14 11-17-55](https://user-images.githubusercontent.com/47980452/62996735-63743a00-be85-11e9-878a-5af9b3301f88.png)
Hello @lahsuk , I am extremely happy that you reached me to help me with this issue. I have completed all the procedures mentioned above and tried testing on messenger. But, following SSL certification error was received. 
![Screenshot from 2019-08-14 11-09-47](https://user-images.githubusercontent.com/47980452/62996657-2b6cf700-be85-11e9-9e8e-e14d11504661.png)
Below are the screenshots of the procedures I followed as mentioned by you
![Screenshot from 2019-08-14 11-11-10](https://user-images.githubusercontent.com/47980452/62996694-417ab780-be85-11e9-908c-5a512c61a756.png)
![Screenshot from 2019-08-14 11-10-34](https://user-images.githubusercontent.com/47980452/62996701-45a6d500-be85-11e9-9cd4-dd67012076b0.png)
![Screenshot from 2019-08-14 11-10-04](https://user-images.githubusercontent.com/47980452/62996708-4b9cb600-be85-11e9-8c8c-840324402b74.png)
Still, thanks for helping me out in this issue. It's been weeks I am unable to solve this issue. If you have time, you can try on this chatbot cloning from this repo. https://gitlab.com/shusrulz/vyccu-bot.git

**Comment by lahsuk** [comment]
@pudasainishushant 

I cloned and tried it and it works fine on my side. You could try re-creating the virtual environment. Maybe the problem is on ngrok's side (I noticed that you're using free tier, while I used pro tier; so maybe...). I can't say anything for certain...

**Comment by abhi99bh** [comment]
@pudasainishushant Hi, were you able to figure out the issue? I am facing the exact same issue. I get the proper response in Postman however in the debug for messenger i get "rasa.core.channels.facebook  - Wrong fb secret! Make sure this matches the secret in your facebook app settings"

**Comment by harishruparel** [comment]
Same issue faced. It worked for the first time but then it just stopped.
Retried the same steps but no success.. Help.

**Comment by tmbo** [comment]
I'll close this as this does not seem related to Rasa, but rather the used SSL certificate (e.g. double check that the root certificate works properly, e.g. using requests in a python shell sending a GET request to facebook's HTTPS URLs should work fine).

In any case, this is more a question than a reproducible bug.
